### ISSUE TITLE
[HOLD setSelection refactor] [$2000] Android - Cursor is placed in wrong position after write/paste the emoji between the word Hello and nice - Reported by: @AndreasBBS

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#12854]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Issue found when executing PR** https://github.com/Expensify/App/pull/12632

## Action Performed:
1. Launch the app
2. Log in with any account
3. Go to any chat
4. Start with a string Hello nice to meet you
5. Write the emoji :wave: in the string Hello nice to meet you between the word Hello and nice
6. Start with a string Hello nice to meet you
7. Paste the string :wave: friend in the string Hello nice to meet you between the word Hello and nice

## Expected Result:
Describe what you think should've happened

## Actual Result:
The resulting string has the emoji converted and the cursor is placed after the word friend/ or after the emoji

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Android

**Version Number:** 1.2.29.2

**Reproducible in staging?:** Yes

**Reproducible in production?:** No

**Email or phone of affected tester (no customers):** 

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/202586022-096c2294-9027-4a31-baf7-d96fc7c5dbec.mp4

**Expensify/Expensify Issue URL:** 

**Issue reported by:** Applause - Internal Team

**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0134ac93ef23b19fbf</li>
        <li>Upwork Job ID: 1594665437635485696</li>
        <li>Last Price Increase: 2022-12-16</li>
    </ul>
</details>

**Comment by roryabraham** [comment]
Can we please add some automated UI tests to cover this behavior with the next PR to fix this?

**Comment by AndreasBBS** [comment]
@roryabraham You can check my analysis of this problem [here](https://github.com/Expensify/App/pull/12632#issuecomment-1312068195). In my opinion is something that needs to be solved at the level of the react-native library, the onSelectionChange has erratic behavior on Android.

**Comment by dylanexpensify** [comment]
reviewing today!

**Comment by dylanexpensify** [comment]
@AndreasBBS added you to title for reporting. 

Upwork job: https://www.upwork.com/jobs/~013f07f27493ef4401

**Comment by Puneet-here** [comment]
Hi, I had reported this bug [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1667850534873129).

**Comment by s77rt** [comment]
### Proposal
```diff
diff --git a/src/components/Composer/index.android.js b/src/components/Composer/index.android.js
index 603bdf829..83dc6be58 100644
--- a/src/components/Composer/index.android.js
+++ b/src/components/Composer/index.android.js
@@ -24,12 +24,6 @@ const propTypes = {
     /** Prevent edits and interactions like focus for this input. */
     isDisabled: PropTypes.bool,
 
-    /** Selection Object */
-    selection: PropTypes.shape({
-        start: PropTypes.number,
-        end: PropTypes.number,
-    }),
-
     /** Whether the full composer can be opened */
     isFullComposerAvailable: PropTypes.bool,
 
@@ -51,10 +45,6 @@ const defaultProps = {
     autoFocus: false,
     isDisabled: false,
     forwardedRef: null,
-    selection: {
-        start: 0,
-        end: 0,
-    },
     isFullComposerAvailable: false,
     setIsFullComposerAvailable: () => {},
     isComposerFullSize: false,
diff --git a/src/components/Composer/index.ios.js b/src/components/Composer/index.ios.js
index b31a5b462..ee3abf32b 100644
--- a/src/components/Composer/index.ios.js
+++ b/src/components/Composer/index.ios.js
@@ -24,12 +24,6 @@ const propTypes = {
     /** Prevent edits and interactions like focus for this input. */
     isDisabled: PropTypes.bool,
 
-    /** Selection Object */
-    selection: PropTypes.shape({
-        start: PropTypes.number,
-        end: PropTypes.number,
-    }),
-
     /** Whether the full composer can be opened */
     isFullComposerAvailable: PropTypes.bool,
 
@@ -51,10 +45,6 @@ const defaultProps = {
     autoFocus: false,
     isDisabled: false,
     forwardedRef: null,
-    selection: {
-        start: 0,
-        end: 0,
-    },
     isFullComposerAvailable: false,
     setIsFullComposerAvailable: () => {},
     isComposerFullSize: false,
@@ -94,9 +84,6 @@ class Composer extends React.Component {
     render() {
         // On native layers we like to have the Text Input not focused so the
         // user can read new chats without the keyboard in the way of the view.
-        // On Android, the selection prop is required on the TextInput but this prop has issues on IOS
-        // https://github.com/facebook/react-native/issues/29063
-        const propsToPass = _.omit(this.props, 'selection');
         return (
             <RNTextInput
                 autoComplete="off"
@@ -108,7 +95,7 @@ class Composer extends React.Component {
                 textAlignVertical="center"
                 style={this.state.propStyles}
                 /* eslint-disable-next-line react/jsx-props-no-spreading */
-                {...propsToPass}
+                {...this.props}
                 editable={!this.props.isDisabled}
             />
         );
diff --git a/src/pages/home/report/ReportActionCompose.js b/src/pages/home/report/ReportActionCompose.js
index 2335864f9..fbc7ca93a 100644
--- a/src/pages/home/report/ReportActionCompose.js
+++ b/src/pages/home/report/ReportActionCompose.js
@@ -47,6 +47,7 @@ import ExceededCommentLength from '../../../components/ExceededCommentLength';
 import withNavigationFocus from '../../../components/withNavigationFocus';
 import * as EmojiUtils from '../../../libs/EmojiUtils';
 import reportPropTypes from '../../reportPropTypes';
+import setSelection from '../../../libs/setSelection';
 
 const propTypes = {
     /** Beta features list */
@@ -323,13 +324,11 @@ class ReportActionCompose extends React.Component {
         const emojiWithSpace = `${emoji} `;
         const newComment = this.comment.slice(0, this.state.selection.start)
             + emojiWithSpace + this.comment.slice(this.state.selection.end, this.comment.length);
-        this.setState(prevState => ({
-            selection: {
-                start: prevState.selection.start + emojiWithSpace.length,
-                end: prevState.selection.start + emojiWithSpace.length,
-            },
-        }));
-        this.updateComment(newComment);
+        const newSelection = {
+            start: this.state.selection.start + emojiWithSpace.length,
+            end: this.state.selection.start + emojiWithSpace.length,
+        };
+        this.updateComment(newComment, false, newSelection);
     }
 
     /**
@@ -381,21 +380,19 @@ class ReportActionCompose extends React.Component {
      * @param {String} comment
      * @param {Boolean} shouldDebounceSaveComment
      */
-    updateComment(comment, shouldDebounceSaveComment) {
+    updateComment(comment, shouldDebounceSaveComment, selection) {
+        const oldComment = this.state.value;
         const newComment = EmojiUtils.replaceEmojis(comment);
-        this.setState((prevState) => {
-            const newState = {
-                isCommentEmpty: !!newComment.match(/^(\s|`)*$/),
-                value: newComment,
-            };
-            if (comment !== newComment) {
-                const remainder = prevState.value.slice(prevState.selection.end).length;
-                newState.selection = {
-                    start: newComment.length - remainder,
-                    end: newComment.length - remainder,
-                };
+        this.setState({
+            isCommentEmpty: !!newComment.match(/^(\s|`)*$/),
+            value: newComment,
+        }, () => {
+            if (selection) {
+                setSelection(this.textInput, selection.start, selection.end);
+            } else if (newComment !== comment) {
+                const remainder = oldComment.slice(this.state.selection.end).length;
+                setSelection(this.textInput, newComment.length - remainder, newComment.length - remainder);
             }
-            return newState;
         });
 
         // Indicate that draft has been created.
@@ -669,7 +666,6 @@ class ReportActionCompose extends React.Component {
                                         shouldClear={this.state.textInputShouldClear}
                                         onClear={() => this.setTextInputShouldClear(false)}
                                         isDisabled={isComposeDisabled || isBlockedFromConcierge}
-                                        selection={this.state.selection}
                                         onSelectionChange={this.onSelectionChange}
                                         isFullComposerAvailable={this.state.isFullComposerAvailable}
                                         setIsFullComposerAvailable={this.setIsFullComposerAvailable}
diff --git a/src/pages/home/report/ReportActionItemMessageEdit.js b/src/pages/home/report/ReportActionItemMessageEdit.js
index 2f7fbc29d..92a53bee8 100644
--- a/src/pages/home/report/ReportActionItemMessageEdit.js
+++ b/src/pages/home/report/ReportActionItemMessageEdit.js
@@ -22,6 +22,7 @@ import * as EmojiUtils from '../../../libs/EmojiUtils';
 import reportPropTypes from '../../reportPropTypes';
 import ExceededCommentLength from '../../../components/ExceededCommentLength';
 import CONST from '../../../CONST';
+import setSelection from '../../../libs/setSelection';
 
 const propTypes = {
     /** All the data of the action */
@@ -100,18 +101,18 @@ class ReportActionItemMessageEdit extends React.Component {
      *
      * @param {String} draft
      */
-    updateDraft(draft) {
+    updateDraft(draft, selection) {
+        const oldDraft = this.state.draft;
         const newDraft = EmojiUtils.replaceEmojis(draft);
-        this.setState((prevState) => {
-            const newState = {draft: newDraft};
-            if (draft !== newDraft) {
-                const remainder = prevState.draft.slice(prevState.selection.end).length;
-                newState.selection = {
-                    start: newDraft.length - remainder,
-                    end: newDraft.length - remainder,
-                };
+        this.setState({
+            draft: newDraft,
+        }, () => {
+            if (selection) {
+                setSelection(this.textInput, selection.start, selection.end);
+            } else if (newDraft !== draft) {
+                const remainder = oldDraft.slice(this.state.selection.end).length;
+                setSelection(this.textInput, newDraft.length - remainder, newDraft.length - remainder);
             }
-            return newState;
         });
 
         // This component is rendered only when draft is set to a non-empty string. In order to prevent component
@@ -180,13 +181,11 @@ class ReportActionItemMessageEdit extends React.Component {
         const emojiWithSpace = `${emoji} `;
         const newComment = this.state.draft.slice(0, this.state.selection.start)
             + emojiWithSpace + this.state.draft.slice(this.state.selection.end, this.state.draft.length);
-        this.setState(prevState => ({
-            selection: {
-                start: prevState.selection.start + emojiWithSpace.length,
-                end: prevState.selection.start + emojiWithSpace.length,
-            },
-        }));
-        this.updateDraft(newComment);
+        const newSelection = {
+            start: this.state.selection.start + emojiWithSpace.length,
+            end: this.state.selection.start + emojiWithSpace.length,
+        };
+        this.updateDraft(newComment, newSelection);
     }
 
     /**
@@ -243,7 +242,6 @@ class ReportActionItemMessageEdit extends React.Component {
                             this.setState({isFocused: false});
                             toggleReportActionComposeView(true, VirtualKeyboard.shouldAssumeIsOpen());
                         }}
-                        selection={this.state.selection}
                         onSelectionChange={this.onSelectionChange}
                     />
                     <View style={styles.editChatItemEmojiWrapper}>
```

### Details
1. Removed `selection` prop since it's not needed and it's also the main cause. We don't need to constantly control the selection just on updateComment, and in my proposal the update is done via `setSelection` lib.

**Comment by mollfpr** [comment]
@s77rt Thanks for the proposal!

Could you attach the result of your proposal?

**Comment by s77rt** [comment]
@mollfpr 

https://user-images.githubusercontent.com/16493223/203107825-612c6615-455d-427a-9059-6e9f537418ff.mp4

**Comment by mollfpr** [comment]
I’ll start reviewing it in the morning, thanks!

**Comment by grgia** [comment]
@mollfpr I believe this was fixed by https://github.com/Expensify/App/pull/12632

**Comment by mollfpr** [comment]
Thanks, @grgia!

This issue is from that PR. You can check the analysis here https://github.com/Expensify/App/issues/12854#issuecomment-1319571877

**Comment by mollfpr** [comment]
Thanks for the reminder! We are still in the middle of accepting proposals here.

**Comment by mollfpr** [comment]
Need more time to review [this](https://github.com/Expensify/App/issues/12854#issuecomment-1322279182)...

**Comment by Pujan92** [comment]
**Proposal**

Find the last index of the unmatched character in comment and newComment string, that index should be the selectionIndex/cursor index.

https://github.com/Expensify/App/blob/532787af6185b463ae040ae3429fa54cb00d21c0/src/pages/home/report/ReportActionCompose.js#L391-L397

```
if (comment !== newComment) {
   const index = this.getLastUnmatchedIndex(comment, newComment);
   newState.selection = {
        start: index,
        end: index,
    };
}
```
New Function
```
   getLastUnmatchedIndex(text, modifiedText) {
        for (let i = text.length - 1, j = modifiedText.length - 1; i >= 0; i--, j--) {
            if (text[i] !== modifiedText[j]) {
                return j;
            }
        }
    }
```

Working Example
[43a2604c-8abf-4441-ac7d-c36cfbd13d22.webm](https://user-images.githubusercontent.com/14358475/203343818-87c5a25c-cf8d-4d90-be46-9854aa791ea6.webm)

**Comment by Karim-30** [comment]
### Proposal

I'm so sorry for this bug, I know it should be covered in my earlier PR.

Anyway, we should make the [replaceEmojis](https://github.com/Expensify/App/blob/0efa4206d8000e6a94423cf10ac4bcf2c44c034b/src/libs/EmojiUtils.js#L191-L211) function smarter enough to determine the writing cursor position not only replacing the `:emoji_name:` with its emoji, so it should return the `newComment` + `cursorPosition`.

**Code :** 

In `/src/libs/EmojiUtils.js` :
```javascript
/**
 * Replace any emoji name in a text with the emoji icon
 * @param {String} text
 * @param {Object} selection
 * @returns {Object}
 */
function replaceEmojis(text, selection) {
    let newText = text;
    let cursorPosition;
    const emojiData = text.match(CONST.REGEX.EMOJI_NAME);
    if (!emojiData || emojiData.length === 0) {
        return {newComment: text, selection};
    }
    for (let i = 0; i < emojiData.length; i++) {
        const checkEmoji = emojisTrie.search(emojiData[i].slice(1, -1));
        if (checkEmoji && checkEmoji.metaData.code) {
            cursorPosition = newText.search(emojiData[i]) + 2;
            newText = newText.replace(emojiData[i], checkEmoji.metaData.code);
        }
    }
    return {newComment: newText, selection: {start: cursorPosition, end: cursorPosition}};
}

```

In `/src/pages/home/report/ReportActionCompose.js` :

```javascript
    updateComment(comment, shouldDebounceSaveComment) {
        const {newComment, selection} = EmojiUtils.replaceEmojis(comment, this.state.selection);
        this.setState({
            isCommentEmpty: !!newComment.match(/^(\s|`)*$/),
            value: newComment,
            selection,
        });
```

in `/src/pages/home/report/ReportActionItemMessageEdit.js` : 
```javascript
    updateDraft(draft) {
        const {newComment: newDraft, selection} = EmojiUtils.replaceEmojis(draft, this.state.selection);
        this.setState({
            draft: newDraft,
            selection,
        });
```

**Result:**


https://user-images.githubusercontent.com/108357004/203421699-96a60a85-a8ea-49d8-b3e8-f37a39fb6272.mov

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by AndreasBBS** [comment]
@Karim-30 @Pujan92 This issue was solved in all platforms except Android. I don't think the proposed solutions will address the problem in Android but if you think it does please post a video of the test in the Android app which is the only platform where this bug is present.

@s77rt approach seems to me the correct one and the one I advised for. Removing the selection prop avoids the onSelectionChange erratic behavior. I'd like to ask you if there's a specific reason you kept the onSelectionChange functionality on your proposal?

**Comment by AndreasBBS** [comment]
@s77rt Yeah you're right thanks for clarifying. I like your solution, good job.

**Comment by Pujan92** [comment]
> @Karim-30 @Pujan92 This issue was solved in all platforms except Android. I don't think the proposed solutions will address the problem in Android but if you think it does please post a video of the test in the Android app which is the only platform where this bug is present.

@AndreasBBS I missed reading the platform in the description, my solution isn't for android specific and not working in android too. So that proposal needs to be ignored.

**Comment by Karim-30** [comment]
>This issue was solved in all platforms except Android.

@AndreasBBS Well done. I using an older version of ExpensifyApp so I didn't see the new code.

**Comment by AndreasBBS** [comment]
@Karim-30 

When developing for the Android platform you can run on an emulator with `npm run android`. This will build the App for Android with your current changes, spin up an emulator and get it running in an emulator. Then you need to run `npx react-native start --port=8083` to connect the App that's running in the emulator. It helps if you have Android Studio installed because it installs all the relevant dependencies for the Android Emulators. If you need help message me on slack, U049U570QPM.

**Comment by mollfpr** [comment]
@s77rt I just try your [proposal](https://github.com/Expensify/App/issues/12854#issuecomment-1322279182), the result looks fine on Android, but not on iOS. When typing the `:wave:` emoji between text, the cursor should be placed next to the emoji instead the next letter.

https://user-images.githubusercontent.com/25520267/203613223-47cdc778-9ed1-4f71-b7d5-0545f1cbd4b5.mov

Could you check this and update your proposal?

**Comment by mollfpr** [comment]
@Pujan92 Could you attach the result for the Android and iOS? Thanks!

**Comment by s77rt** [comment]
That's another bug in react-native https://github.com/facebook/react-native/issues/28865
@mollfpr Should we make this platform specific?

**Comment by s77rt** [comment]
@mollfpr If platform-specific component is not needed here; we can on the same file:
1. Use two state values: `selection` and `prevSelection`
2. On `updateComment` we use `this.state.prevSelection` for iOS since the intended selection was already changed and `this.state.selection` for the rest.

on iOS the events callback are invested, the first call is `onSelectionChange` and then `onChangeText`

**Comment by mollfpr** [comment]
Checking again…

**Comment by Pujan92** [comment]
> @Pujan92 Could you attach the result for the Android and iOS? Thanks!

I have a hacky way which isn't the preferred way but it can work for android too. The issue we are facing in the android is `onSelectionChange` call which executes after we set the `selection` state in the `updateComment` which causes infinite cursor flickering/positioning.

**Hacky Solution**
Extra state property `shouldAllowOnSelectionChange`

In update emoji comment update the `shouldAllowOnSelectionChange` property
https://github.com/Expensify/App/blob/8c86a77a69bef6f4939c4642e33ff413ab97be07/src/pages/home/report/ReportActionCompose.js#L391-L397
`newState.shouldAllowOnSelectionChange = false;`

In the case of emoji addition we won't update the `selection` state from `onSelectionChange` and apply the minor delay to avoid flickering.

https://github.com/Expensify/App/blob/8c86a77a69bef6f4939c4642e33ff413ab97be07/src/pages/home/report/ReportActionCompose.js#L196-L198
```
   onSelectionChange(e) {
        if (this.state.shouldAllowOnSelectionChange) {
            this.setState({selection: e.nativeEvent.selection});
        } else {
            setTimeout(() => {
                this.setState({shouldAllowOnSelectionChange: true});
            }, 100);
        }
    }
```

**Comment by mollfpr** [comment]
@s77rt Sorry for the wait. How about we create a function lib similar to `setSelection` to set the selection for composer input? Like `setComposerSelection` and return empty on iOS?

@Pujan92 Thanks for the update! We avoid using `setTimeout` for a solution.

**Comment by s77rt** [comment]
@mollfpr I don't see the point of this? We do need to keep a track of the previous selection
The new selection is based on the old selection.

The platform specific code is only one line maybe we can make an exception for it:
`const caretPosition = Platform.OS === "ios" ? {...this.state.prevSelection} : {...this.state.selection};`

**Comment by mollfpr** [comment]
@s77rt Could you put the solution into the full proposal? I'm still not convinced with the usage of `this.state.prevSelection`.

**Comment by s77rt** [comment]
@mollfpr on iOS the selection on this.state.selection gets updated before calling the updateComment
that's why we need to store another selection (prevSelection)

will post the updated proposal in few mins

**Comment by s77rt** [comment]
### Proposal (Updated)
```diff
diff --git a/src/components/Composer/index.android.js b/src/components/Composer/index.android.js
index 603bdf829e..83dc6be587 100644
--- a/src/components/Composer/index.android.js
+++ b/src/components/Composer/index.android.js
@@ -24,12 +24,6 @@ const propTypes = {
     /** Prevent edits and interactions like focus for this input. */
     isDisabled: PropTypes.bool,
 
-    /** Selection Object */
-    selection: PropTypes.shape({
-        start: PropTypes.number,
-        end: PropTypes.number,
-    }),
-
     /** Whether the full composer can be opened */
     isFullComposerAvailable: PropTypes.bool,
 
@@ -51,10 +45,6 @@ const defaultProps = {
     autoFocus: false,
     isDisabled: false,
     forwardedRef: null,
-    selection: {
-        start: 0,
-        end: 0,
-    },
     isFullComposerAvailable: false,
     setIsFullComposerAvailable: () => {},
     isComposerFullSize: false,
diff --git a/src/components/Composer/index.ios.js b/src/components/Composer/index.ios.js
index b31a5b462f..ee3abf32b3 100644
--- a/src/components/Composer/index.ios.js
+++ b/src/components/Composer/index.ios.js
@@ -24,12 +24,6 @@ const propTypes = {
     /** Prevent edits and interactions like focus for this input. */
     isDisabled: PropTypes.bool,
 
-    /** Selection Object */
-    selection: PropTypes.shape({
-        start: PropTypes.number,
-        end: PropTypes.number,
-    }),
-
     /** Whether the full composer can be opened */
     isFullComposerAvailable: PropTypes.bool,
 
@@ -51,10 +45,6 @@ const defaultProps = {
     autoFocus: false,
     isDisabled: false,
     forwardedRef: null,
-    selection: {
-        start: 0,
-        end: 0,
-    },
     isFullComposerAvailable: false,
     setIsFullComposerAvailable: () => {},
     isComposerFullSize: false,
@@ -94,9 +84,6 @@ class Composer extends React.Component {
     render() {
         // On native layers we like to have the Text Input not focused so the
         // user can read new chats without the keyboard in the way of the view.
-        // On Android, the selection prop is required on the TextInput but this prop has issues on IOS
-        // https://github.com/facebook/react-native/issues/29063
-        const propsToPass = _.omit(this.props, 'selection');
         return (
             <RNTextInput
                 autoComplete="off"
@@ -108,7 +95,7 @@ class Composer extends React.Component {
                 textAlignVertical="center"
                 style={this.state.propStyles}
                 /* eslint-disable-next-line react/jsx-props-no-spreading */
-                {...propsToPass}
+                {...this.props}
                 editable={!this.props.isDisabled}
             />
         );
diff --git a/src/pages/home/report/ReportActionCompose.js b/src/pages/home/report/ReportActionCompose.js
index 2335864f91..90c571e009 100644
--- a/src/pages/home/report/ReportActionCompose.js
+++ b/src/pages/home/report/ReportActionCompose.js
@@ -4,6 +4,7 @@ import {
     View,
     TouchableOpacity,
     InteractionManager,
+    Platform,
 } from 'react-native';
 import _ from 'underscore';
 import lodashGet from 'lodash/get';
@@ -47,6 +48,7 @@ import ExceededCommentLength from '../../../components/ExceededCommentLength';
 import withNavigationFocus from '../../../components/withNavigationFocus';
 import * as EmojiUtils from '../../../libs/EmojiUtils';
 import reportPropTypes from '../../reportPropTypes';
+import setSelection from '../../../libs/setSelection';
 
 const propTypes = {
     /** Beta features list */
@@ -142,6 +144,10 @@ class ReportActionCompose extends React.Component {
                 start: props.comment.length,
                 end: props.comment.length,
             },
+            prevSelection: {
+                start: props.comment.length,
+                end: props.comment.length,
+            },
             maxLines: props.isSmallScreenWidth ? CONST.COMPOSER.MAX_LINES_SMALL_SCREEN : CONST.COMPOSER.MAX_LINES,
             value: props.comment,
 
@@ -194,7 +200,12 @@ class ReportActionCompose extends React.Component {
     }
 
     onSelectionChange(e) {
-        this.setState({selection: e.nativeEvent.selection});
+        this.setState(prevState => {
+            return {
+                selection: e.nativeEvent.selection,
+                prevSelection: {...prevState.selection},
+            };
+        });
     }
 
     /**
@@ -323,13 +334,11 @@ class ReportActionCompose extends React.Component {
         const emojiWithSpace = `${emoji} `;
         const newComment = this.comment.slice(0, this.state.selection.start)
             + emojiWithSpace + this.comment.slice(this.state.selection.end, this.comment.length);
-        this.setState(prevState => ({
-            selection: {
-                start: prevState.selection.start + emojiWithSpace.length,
-                end: prevState.selection.start + emojiWithSpace.length,
-            },
-        }));
-        this.updateComment(newComment);
+        const newSelection = {
+            start: this.state.selection.start + emojiWithSpace.length,
+            end: this.state.selection.start + emojiWithSpace.length,
+        };
+        this.updateComment(newComment, false, newSelection);
     }
 
     /**
@@ -381,21 +390,24 @@ class ReportActionCompose extends React.Component {
      * @param {String} comment
      * @param {Boolean} shouldDebounceSaveComment
      */
-    updateComment(comment, shouldDebounceSaveComment) {
+    updateComment(comment, shouldDebounceSaveComment, selection) {
+        // We require the caret position just before updating the comment
+        // on iOS the position gets updated before calling the onChangeText callback (updateComment)
+        // i.e. onSelectionChange => onChangeText. Thus the intended selection is overwritten already that's why we use prevSelection
+        // on the rest of the platforms the callback flow is onChangeText => onSelectionChange. The intended selection is still saved on state.
+        const caretPosition = Platform.OS === "ios" ? {...this.state.prevSelection} : {...this.state.selection};
+        const oldComment = this.state.value;
         const newComment = EmojiUtils.replaceEmojis(comment);
-        this.setState((prevState) => {
-            const newState = {
-                isCommentEmpty: !!newComment.match(/^(\s|`)*$/),
-                value: newComment,
-            };
-            if (comment !== newComment) {
-                const remainder = prevState.value.slice(prevState.selection.end).length;
-                newState.selection = {
-                    start: newComment.length - remainder,
-                    end: newComment.length - remainder,
-                };
+        this.setState({
+            isCommentEmpty: !!newComment.match(/^(\s|`)*$/),
+            value: newComment,
+        }, () => {
+            if (selection) {
+                setSelection(this.textInput, selection.start, selection.end);
+            } else if (newComment !== comment) {
+                const remainder = oldComment.slice(caretPosition.end).length;
+                setSelection(this.textInput, newComment.length - remainder, newComment.length - remainder);
             }
-            return newState;
         });
 
         // Indicate that draft has been created.
@@ -669,7 +681,6 @@ class ReportActionCompose extends React.Component {
                                         shouldClear={this.state.textInputShouldClear}
                                         onClear={() => this.setTextInputShouldClear(false)}
                                         isDisabled={isComposeDisabled || isBlockedFromConcierge}
-                                        selection={this.state.selection}
                                         onSelectionChange={this.onSelectionChange}
                                         isFullComposerAvailable={this.state.isFullComposerAvailable}
                                         setIsFullComposerAvailable={this.setIsFullComposerAvailable}
diff --git a/src/pages/home/report/ReportActionItemMessageEdit.js b/src/pages/home/report/ReportActionItemMessageEdit.js
index 2f7fbc29db..5a41f6fec0 100644
--- a/src/pages/home/report/ReportActionItemMessageEdit.js
+++ b/src/pages/home/report/ReportActionItemMessageEdit.js
@@ -1,6 +1,6 @@
 import lodashGet from 'lodash/get';
 import React from 'react';
-import {InteractionManager, View} from 'react-native';
+import {InteractionManager, View, Platform} from 'react-native';
 import PropTypes from 'prop-types';
 import _ from 'underscore';
 import ExpensiMark from 'expensify-common/lib/ExpensiMark';
@@ -22,6 +22,7 @@ import * as EmojiUtils from '../../../libs/EmojiUtils';
 import reportPropTypes from '../../reportPropTypes';
 import ExceededCommentLength from '../../../components/ExceededCommentLength';
 import CONST from '../../../CONST';
+import setSelection from '../../../libs/setSelection';
 
 const propTypes = {
     /** All the data of the action */
@@ -82,6 +83,10 @@ class ReportActionItemMessageEdit extends React.Component {
                 start: draftMessage.length,
                 end: draftMessage.length,
             },
+            prevSelection: {
+                start: draftMessage.length,
+                end: draftMessage.length,
+            },
             isFocused: false,
         };
     }
@@ -92,7 +97,12 @@ class ReportActionItemMessageEdit extends React.Component {
      * @param {Event} e
      */
     onSelectionChange(e) {
-        this.setState({selection: e.nativeEvent.selection});
+        this.setState(prevState => {
+            return {
+                selection: e.nativeEvent.selection,
+                prevSelection: {...prevState.selection},
+            };
+        });
     }
 
     /**
@@ -100,18 +110,23 @@ class ReportActionItemMessageEdit extends React.Component {
      *
      * @param {String} draft
      */
-    updateDraft(draft) {
+    updateDraft(draft, selection) {
+        // We require the caret position just before updating the comment
+        // on iOS the position gets updated before calling the onChangeText callback (updateComment)
+        // i.e. onSelectionChange => onChangeText. Thus the intended selection is overwritten already that's why we use prevSelection
+        // on the rest of the platforms the callback flow is onChangeText => onSelectionChange. The intended selection is still saved on state.
+        const caretPosition = Platform.OS === "ios" ? {...this.state.prevSelection} : {...this.state.selection};
+        const oldDraft = this.state.draft;
         const newDraft = EmojiUtils.replaceEmojis(draft);
-        this.setState((prevState) => {
-            const newState = {draft: newDraft};
-            if (draft !== newDraft) {
-                const remainder = prevState.draft.slice(prevState.selection.end).length;
-                newState.selection = {
-                    start: newDraft.length - remainder,
-                    end: newDraft.length - remainder,
-                };
+        this.setState({
+            draft: newDraft,
+        }, () => {
+            if (selection) {
+                setSelection(this.textInput, selection.start, selection.end);
+            } else if (newDraft !== draft) {
+                const remainder = oldDraft.slice(caretPosition.end).length;
+                setSelection(this.textInput, newDraft.length - remainder, newDraft.length - remainder);
             }
-            return newState;
         });
 
         // This component is rendered only when draft is set to a non-empty string. In order to prevent component
@@ -180,13 +195,11 @@ class ReportActionItemMessageEdit extends React.Component {
         const emojiWithSpace = `${emoji} `;
         const newComment = this.state.draft.slice(0, this.state.selection.start)
             + emojiWithSpace + this.state.draft.slice(this.state.selection.end, this.state.draft.length);
-        this.setState(prevState => ({
-            selection: {
-                start: prevState.selection.start + emojiWithSpace.length,
-                end: prevState.selection.start + emojiWithSpace.length,
-            },
-        }));
-        this.updateDraft(newComment);
+        const newSelection = {
+            start: this.state.selection.start + emojiWithSpace.length,
+            end: this.state.selection.start + emojiWithSpace.length,
+        };
+        this.updateDraft(newComment, newSelection);
     }
 
     /**
@@ -243,7 +256,6 @@ class ReportActionItemMessageEdit extends React.Component {
                             this.setState({isFocused: false});
                             toggleReportActionComposeView(true, VirtualKeyboard.shouldAssumeIsOpen());
                         }}
-                        selection={this.state.selection}
                         onSelectionChange={this.onSelectionChange}
                     />
                     <View style={styles.editChatItemEmojiWrapper}>
```

### Update
Support iOS



https://user-images.githubusercontent.com/16493223/204034172-391d43a8-29df-472d-aedc-b751b2efafe9.mp4

**Comment by mollfpr** [comment]
The current proposal from @s77rt is suggesting handling both input selections for iOS and Android. It's required to create 2 states for handling each iOS and Android input selection because this RN issue https://github.com/facebook/react-native/issues/28865.

Should we hold this until the issue on RN is fixed (I doubt it will be fixed in near future) or accepting a workaround for that?

cc @MariaHCD @dylanexpensify

**Comment by dylanexpensify** [comment]
Sorry, @rushatgabhane it looks like @Puneet-here noted he reported this bug earlier than @AndreasBBS did - (the 7th vs the 11th), if so, we'd issue Puneet the reporting bonus. Can you confirm?

**Comment by dylanexpensify** [comment]
@rushatgabhane I'm referring to [this comment](https://github.com/Expensify/App/issues/12854#issuecomment-1319571877)!

**Comment by AndreasBBS** [comment]
> @dylanexpensify thank you. @Puneet-here reported the bug earlier than @AndreasBBS
> 
> @Puneet-here reported the bug on Nov 8 over here https://expensify.slack.com/archives/C01GTK53T8Q/p1667850534873129?thread_ts=1667850534.873129&cid=C01GTK53T8Q

I cannot open this slack archive but I only reported the bug on November 11 so I think @Puneet-here should get this one. Sorry for the confusion, I was not aware of this previous reporting.

**Comment by MariaHCD** [comment]
I'm currently reviewing the proposal! I'll update here shortly.

**Comment by dylanexpensify** [comment]
@MariaHCD can we get an update today? 🙇‍♂️

**Comment by MariaHCD** [comment]
@mollfpr @s77rt We aren't accepting platform-specific workarounds for this issue. Our [App philosophy](https://github.com/Expensify/App#philosophy) (item 5.ii) specifically mentions:

```
If the reason you can't write cross-platform code is because there is a bug in ReactNative that is preventing it from working, the correct action is to fix RN and submit a PR upstream -- not to hack around RN bugs with platform-specific code paths.
```

Looks best way would be to submit a fix to RN.

**Comment by mollfpr** [comment]
Still waiting for a proposal!

**Comment by dylanexpensify** [comment]
to confirm @mollfpr we have no pending proposals, correct?

**Comment by mollfpr** [comment]
@dylanexpensify Yup, we don't.

**Comment by dylanexpensify** [comment]
Great, given it's been some time I'm going to double then

**Comment by dylanexpensify** [comment]
doubled - https://expensify.slack.com/archives/C01GTK53T8Q/p1670527150372249

**Comment by s77rt** [comment]
I basically provided a solution. Do I need to fix this on react-native in order to actually get my proposal selected?
I mean anyone can fix the issue on react-native and copy paste my proposal

**Comment by mollfpr** [comment]
@s77rt Which proposal do you refer to as the fix in the RN upstream?

We can accept your [proposal](https://github.com/Expensify/App/issues/12854#issuecomment-1322279182) but it's before that we need to fix this RN [issue](https://github.com/Expensify/App/issues/12854#issuecomment-1325529776) to make your proposal work both iOS and Android.

**Comment by s77rt** [comment]
@mollfpr I meant a solution to this issue not RN's issue.
What I'm asking is what if another person fixed the RN's issue and copy pasted my proposal above

Does he get all the credit? Or is that a situation where we split the compensation?

**Comment by mollfpr** [comment]
@s77rt Good question! 

@MariaHCD how do we handle this kind of situation?

**Comment by s77rt** [comment]
I took a look into this and I can say that root cause is not in `react-native` but in iOS it self.
iOS is fires [textViewDidChangeSelection](https://developer.apple.com/documentation/uikit/uitextviewdelegate/1618620-textviewdidchangeselection?language=objc) then [textViewDidChange](https://developer.apple.com/documentation/uikit/uitextviewdelegate/1618599-textviewdidchange?language=objc)

Meaning, even a solution at `react-native` level would have to be a hacky solution

**Comment by s77rt** [comment]
Quick update!

I have found that react team actually did fix this on [react fabric](https://github.com/facebook/react-native/blob/main/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm#L36-L46)

But it's not fixed on react native, the fix is hacky as expected but since I'm just re-using their code and there is no other way around it to fix this, a PR may be accepted.

So here is the fix:
```diff
diff --git a/RCTBackedTextInputDelegateAdapter.m b/RCTBackedTextInputDelegateAdapter2.m
index c6c254c..ebbb4f6 100644
--- a/RCTBackedTextInputDelegateAdapter.m
+++ b/RCTBackedTextInputDelegateAdapter2.m
@@ -162,6 +162,8 @@ static void *TextFieldSelectionObservingContext = &TextFieldSelectionObservingCo
 
 @implementation RCTBackedTextViewDelegateAdapter {
   __weak UITextView<RCTBackedTextInputViewProtocol> *_backedTextInputView;
+  NSAttributedString *_lastStringStateWasUpdatedWith;
+  BOOL _ignoreNextTextInputCall;
   BOOL _textDidChangeIsComing;
   UITextRange *_previousSelectedTextRange;
 }
@@ -246,12 +248,21 @@ static void *TextFieldSelectionObservingContext = &TextFieldSelectionObservingCo
 
 - (void)textViewDidChange:(__unused UITextView *)textView
 {
+  if (_ignoreNextTextInputCall && [_lastStringStateWasUpdatedWith isEqual:_backedTextInputView.attributedText]) {
+    _ignoreNextTextInputCall = NO;
+    return;
+  }
+  _lastStringStateWasUpdatedWith = _backedTextInputView.attributedText;
   _textDidChangeIsComing = NO;
   [_backedTextInputView.textInputDelegate textInputDidChange];
 }
 
 - (void)textViewDidChangeSelection:(__unused UITextView *)textView
 {
+  if (![_lastStringStateWasUpdatedWith isEqual:_backedTextInputView.attributedText]) {
+    [self textViewDidChange:_backedTextInputView];
+    _ignoreNextTextInputCall = YES;
+  }
   [self textViewProbablyDidChangeSelection];
 }
```
https://github.com/Expensify/react-native/blob/Expensify-0.70.4-alpha.2/Libraries/Text/TextInput/RCTBackedTextInputDelegateAdapter.m

I have tested the code for the events and the order is now as expected `onChangeText` then `onSelectionChange`

**Comment by s77rt** [comment]
We have a PR https://github.com/facebook/react-native/pull/35603

**Comment by MariaHCD** [comment]
Nice, thanks for the update @s77rt!

> What I'm asking is what if another person fixed the RN's issue and copy pasted my proposal above

Does he get all the credit? Or is that a situation where we split the compensation?

Correct me if I'm wrong @dylanexpensify but I think in these cases we generally try to keep one contributor for both the RN issue and our App issue.

In terms of going forward, I'm thinking we should wait for perhaps 1 week for the RN PR (although it will likely take more than that). If there is no reviews and no progress, we can have the fix as part of our RN fork instead.

cc: @roryabraham @AndrewGable

**Comment by mollfpr** [comment]
@MariaHCD Just to clarify, we start to wait for the RN PR https://github.com/facebook/react-native/pull/35603 to move forward then test again the current proposal from @s77rt?

**Comment by MariaHCD** [comment]
@mollfpr I'm thinking we wait for perhaps a week for the RN PR https://github.com/facebook/react-native/pull/35603, if there is no progress there, then we will create the same PR on our fork of RN.

**Comment by dylanexpensify** [comment]
Great question @MariaHCD! I think we generally keep it to one contributor, but help me understand why there'd be two different cases (1 app 1 RN) if new dot is in RN?

**Comment by s77rt** [comment]
@dylanexpensify In this issue we are trying to fix two bugs:
1. The main bug as in the title
2. The RN's bug: inverted events order in iOS

If UserA fixed bug 1 then UserB fixed bug 2

It wouldn't be fair to give all the credit to UserB

**Comment by s77rt** [comment]
The RN PR has been merged

**Comment by MariaHCD** [comment]
Just an update @mollfpr @s77rt, we're working on getting our fork of RN updated so it includes @s77rt's fix: https://github.com/facebook/react-native/pull/35603

In the meantime, @s77rt can you clarify if your [latest proposal](https://github.com/Expensify/App/issues/12854#issuecomment-1327740844) to fix the App issue remains unchanged? If not, could you post an updated proposal?

**Comment by s77rt** [comment]
@MariaHCD We should go with my [first proposal](https://github.com/Expensify/App/issues/12854#issuecomment-1322279182)
The latest proposal was introduced to fix the iOS issue by the use of platform specific logic which is now no longer needed.

The App code has changed a bit, but the proposal will be almost the same. Will post an updated proposal when the RN fork is updated.

**Comment by MariaHCD** [comment]
As an update, we're going to wait until the [merged RN PR](https://github.com/facebook/react-native/pull/35603) is included in a [release](https://github.com/facebook/react-native/releases) before updating our fork. I'll keep an eye out for that release.

**Comment by s77rt** [comment]
@MariaHCD Why I have been assigned early? Won't that effect the PR bonus?

**Comment by MariaHCD** [comment]
> Why I have been assigned early?

Sorry for not clarifying. We want to mark this issue as no longer open for proposals from other contributors. Especially since you've already posted a proposal that needs a bit of tweaking once our RN fork is updated.

> Won't that effect the PR bonus?

In this case, I think the PR bonus will still be applicable. It will just be based on the day when our RN fork is updated and your updated proposal is approved by both @mollfpr and myself. cc: @dylanexpensify

**Comment by mollfpr** [comment]
Sounds good to me 👍

**Comment by dylanexpensify** [comment]
Correct, PR bonus can still be applicable!

**Comment by MariaHCD** [comment]
Not overdue, still waiting for the RN PR to be included in a release.

**Comment by dylanexpensify** [comment]
Any updates on the RN PR being included?

**Comment by mollfpr** [comment]
I’m not familiar with the RN release process 😅

Maybe @s77rt knows?

**Comment by s77rt** [comment]
No idea either.

**Comment by mollfpr** [comment]
I think we should wait for the commit to get picked for the next release. Currently, I don't see the commit picked yet https://github.com/reactwg/react-native-releases/discussions/41.

**Comment by trjExpensify** [comment]
Yup! Updated the issue title to add clarity on what we're held on in the meantime. Let's make sure to keep checking back in for the release 👍

**Comment by roryabraham** [comment]
Posting here for visibility, but since we use a fork of React Native, we'll actually need to integrate the upstream change into our fork and then re-release the fork.

**Comment by MariaHCD** [comment]
Doesn't look like the [RN PR](https://github.com/facebook/react-native/commit/64475aeb3b1b0e37bdad2c2d5abb2116b2141c06) has been picked for the latest release [0.71.0](https://github.com/reactwg/react-native-releases/discussions/41)

https://reactnative.dev/contributing/release-faq#how-do-i-know-if-my-fixfeature-is-in-a-certain-release

```
If the commit is only present in main (i.e. has no tags), then the commit has yet to be picked up by a release (or it may have been included in a follow up cherry pick for a patch version). You can expect it to be included in the next release candidate that is cut once the designed features have all landed.
```

According to the docs, it may take one or two months for the change to make it into a stable React Native release.

https://reactnative.dev/contributing/release-faq#when-will-my-fix-make-it-into-a-release

**Comment by s77rt** [comment]
Maybe we can get it up merged quickly upon request https://github.com/reactwg/react-native-releases/discussions/41#discussioncomment-4548354

**Comment by trjExpensify** [comment]
^^ Nice! So it has been added, right?

**Comment by mollfpr** [comment]
Yes!

**Comment by s77rt** [comment]
Looks like RN 0.71.0 has been just released.

**Comment by mollfpr** [comment]
@s77rt Can you confirm that the change is there and update your proposal?

I see that the commit is in the tag, but I just to double-check it.
<img width="813" alt="Screen Shot 2023-01-13 at 00 20 15" src="https://user-images.githubusercontent.com/25520267/212135669-a3db019f-f4be-4732-be55-378953bd17a9.png">

**Comment by s77rt** [comment]
@mollfpr Yes confirmed it's there.
I will update my proposal once we update to the latest version so we can test reliability.

cc @MariaHCD

**Comment by JmillsExpensify** [comment]
The latest RN version is released right? If so, let's stick the `daily` label back on this issue.

**Comment by MariaHCD** [comment]
The next step would be to have our RN fork updated. Asked In Slack [here](https://expensify.slack.com/archives/C07J32337/p1673944192987039?thread_ts=1671097765.803029&cid=C07J32337).

**Comment by MariaHCD** [comment]
Looks like we're waiting on updating our RN fork: https://expensify.slack.com/archives/C07J32337/p1673977754037079?thread_ts=1671097765.803029&cid=C07J32337

cc: @tgolen

**Comment by dylanexpensify** [comment]
not overdue

**Comment by dylanexpensify** [comment]
@MariaHCD still waiting, yeah?

**Comment by MariaHCD** [comment]
Yes, I believe we're still waiting. Bumped the thread in Slack: https://expensify.slack.com/archives/C07J32337/p1675079403923329?thread_ts=1671097765.803029&cid=C07J32337

**Comment by MariaHCD** [comment]
We should hopefully get our fork updated this week 🤞🏼 

https://expensify.slack.com/archives/C07J32337/p1675080097676869?thread_ts=1671097765.803029&cid=C07J32337

**Comment by dylanexpensify** [comment]
Should we move this to weekly? @MariaHCD

**Comment by dylanexpensify** [comment]
Not overdue - still on hold

**Comment by dylanexpensify** [comment]
not overdue

**Comment by dylanexpensify** [comment]
Not overdue

**Comment by s77rt** [comment]
Not overdue. Still holding for [Updating our RN fork](https://github.com/Expensify/App/issues/15124).

**Comment by s77rt** [comment]
Same ^

**Comment by dylanexpensify** [comment]
sa-sa-sa-same

**Comment by MariaHCD** [comment]
Still on hold for https://github.com/Expensify/App/issues/15124

**Comment by dylanexpensify** [comment]
Still on hold

**Comment by s77rt** [comment]
Cool! It has been so long, hopefully not so much has changed :sweat_smile:.
The bug still exists and I think [the proposal](https://github.com/Expensify/App/issues/12854#issuecomment-1322279182) will pretty much be the same. Will test and provide an update soon.

**Comment by s77rt** [comment]
I confirmed the proposal still stands and fixes the bug. Posting an updated proposal in the next comment.

**Comment by s77rt** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Typing emojis causes the cursor to jump around multiple times and lands in a wrong final position.

### What is the root cause of that problem?
Selection-controlled TextInput on RN is so buggy especially for Android.

### What changes do you think we should make in order to solve the problem?
Instead of controlling the selection the traditional way (passing `selection` prop) which is always render dependent. We can control the selection directly using an independent and render free method which is `setSelection` for Native and `setSelectionRange` for Web. We can use our [setSelection](https://github.com/Expensify/App/tree/main/src/libs/setSelection) library for that.



### What alternative solutions did you explore? (Optional)
I didn't explore this solution but we can always fix the buggy behavior on RN or wait for a fix to land there.

**Comment by s77rt** [comment]
We have this PR https://github.com/Expensify/react-native/pull/50 that may fix the RN bug. Waiting for @hannojg confirmation on this.

**Comment by dylanexpensify** [comment]
Not overdue! @s77rt are we still waiting for @hannojg to confirm?

**Comment by hannojg** [comment]
> @dylanexpensify Actually the PR may no longer be valid for the regressions that it introduces.

That's correct!

> @hannojg I think you were planning to use setSelection at some point. We may go with that, do you have any concerns for that matter? Or are we expecting an update in the PR that we should hold this issue for.

Yes, this is the route to go then. [There is this PR](https://github.com/Expensify/App/pull/11684), which I am working on that uses setSelection, but is currently on hold due to function component refactor. But I assume that would fix this issue as well.

**Comment by s77rt** [comment]
@hannojg Cool, so do you think we can go in parallel here? Check [diff12854.txt](https://github.com/Expensify/App/files/11265960/diff12854.txt) Or would that contradicts with your work?

**Comment by hannojg** [comment]
@s77rt Can't test the diff with latest stage from main, due to this error:

```
error: src/pages/home/report/ReportActionCompose/index.js: No such file or directory
```

but yes, you can go ahead and make that change anyways 😊 I believe it would fix the [android selection bug](https://github.com/Expensify/App/issues/8349), would be interesting to test

**Comment by s77rt** [comment]
@hannojg Yeah, the diff is a few days old but basically the same idea remains. Will raise a PR soon.

**Comment by s77rt** [comment]
I have raised the PR here https://github.com/Expensify/App/pull/17687 cc @MariaHCD @mollfpr @hannojg 
I didn't apply on Upwork as the job is no longer available cc @dylanexpensify 

And I have started a Slack discussion here https://expensify.slack.com/archives/C01GTK53T8Q/p1681938841671029 to be sure we are aligned on that way gong forward.

**Comment by MelvinBot** [comment]
@MariaHCD, @s77rt, @mollfpr, @dylanexpensify Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by s77rt** [comment]
Dear Melvin,

I'm writing to inform you that we will be chasing an upstream fix for the buggy selection prop on RN for both Android and iOS. There should be a P/S statement posted on Slack soon. We will probably create two new issues on E/App and then put this one on hold.

I hope that answers your concerns.

Sincerely yours,
s77rt

**Comment by mountiny** [comment]
I am putting this on hold for the discussion to have, it will be a bigger lift for sure so making this weekly too

**Comment by JmillsExpensify** [comment]
Bahaha, love it. 

> Dear Melvin,

> I'm writing to inform you that we will be chasing an upstream fix for the buggy selection prop on RN for both Android and iOS. There should be a P/S statement posted on Slack soon. We will probably create two new issues on E/App and then put this one on hold.

> I hope that answers your concerns.

> Sincerely yours,
> s77rt

**Comment by hannojg** [comment]
Still looking for a solid fix!

**Comment by dylanexpensify** [comment]
@s77rt will get new job posted today!

**Comment by s77rt** [comment]
Still on hold

**Comment by dylanexpensify** [comment]
@s77rt any update here? Or any estimate when this will come off hold in your opinion? Thanks!

**Comment by mountiny** [comment]
Asking @hannojg whats their plan on this one I think we can make this monthly

**Comment by s77rt** [comment]
They have an upstream PR https://github.com/facebook/react-native/pull/37424 related to text selection which is targeting https://github.com/Expensify/App/issues/8349. However I'm not sure if it will fix this one as well.

**Comment by mountiny** [comment]
I think that came a before but maybe it will be helpful

**Comment by hannojg** [comment]
Hey, I just tested it and can confirm that this bug gets fixed with:

- https://github.com/facebook/react-native/pull/37424

(Note: There are other issues such as https://github.com/Expensify/App/issues/16696 that are on hold for this issue. However, the named PR doesn't fix those issues (the PR fixes a specific android issue, and the other issues are about iOS))

**Comment by mountiny** [comment]
thanks for the update 🎉

**Comment by MariaHCD** [comment]
Thanks, @hannojg!

So in terms of next steps, I believe we're going to have to wait until that PR is part of our RN fork and then retest this.

If it does fix this issue, for payment, I know both @s77rt and @mollfpr put time and effort into this issue. So I think we should compensate both. What do you both think is fair compensation? 

cc: @dylanexpensify

**Comment by dylanexpensify** [comment]
Sounds great! @s77rt and @mollfpr can you reply here please?

**Comment by s77rt** [comment]
@dylanexpensify Yes!

**Comment by dylanexpensify** [comment]
Mind giving us an idea on what seems fair for this, friends? 🙇‍♂️ @s77rt @mollfpr

**Comment by s77rt** [comment]
@mollfpr The bug seems to be fixed by https://github.com/facebook/react-native/pull/37424 https://github.com/Expensify/App/issues/12854#issuecomment-1570047376

@dylanexpensify Since no PR ended up being merged I think 25% is fair for my case.

**Comment by dylanexpensify** [comment]
got it! @s77rt mind bringing this up please to C+ channel and cc'ing me?

**Comment by hannojg** [comment]
@s77rt did you had any chance to double check? I think this is fixed and can get closed 😊

**Comment by hannojg** [comment]
@dylanexpensify I think we can then close this issue, as it was fixed as part of our work to fix the selection prop in react native:

- https://github.com/facebook/react-native/pull/37424

**Comment by MariaHCD** [comment]
Oh, missed the latest updates on this but it sounds like we are good to close!

**Comment by dylanexpensify** [comment]
@MariaHCD mind confirming thoughts on above compensation?

**Comment by MariaHCD** [comment]
Ah, apologies, I missed that we need to pay out the compensations as I suggested [before](https://github.com/Expensify/App/issues/12854#issuecomment-1573851246).

After refreshing my memory on the work on this issue:
- @mollfpr handled a lot of investigation, reviewing proposals and testing.
- @s77rt also did spend time in investigation and worked on a RN PR: https://github.com/facebook/react-native/pull/35603

So the compensation that both suggested sounds good to me too!

**Comment by dylanexpensify** [comment]
Payment Summary:
- Reporter: @AndreasBBS $250
- Contributor: @mollfpr $1,000
- Contributor+: @s77rt $500

Upwork job [here](https://www.upwork.com/jobs/~0119c27c3e9f0700ad)! Please apply!

**Comment by s77rt** [comment]
@dylanexpensify Applied! Thanks!

**Comment by mollfpr** [comment]
Applied too, thanks @dylanexpensify

**Comment by dylanexpensify** [comment]
@s77rt @mollfpr sent offers!

**Comment by mollfpr** [comment]
Accepted, thank you!

**Comment by s77rt** [comment]
Same ^

**Comment by Puneet-here** [comment]
Hey @dylanexpensify, I am the original reporter of this issue. It was confirmed [here](https://github.com/Expensify/App/issues/12854#issuecomment-1330678028) earlier.

**Comment by dylanexpensify** [comment]
@Puneet-here offer sent!
@mollfpr @s77rt payments sent!

**Comment by Puneet-here** [comment]
> @Puneet-here offer sent! 

Accepted!

**Comment by dylanexpensify** [comment]
We're all done! Thanks y'all!
